### PR TITLE
Fixed AmbiguousMatchException in DataAnnotationsValidator for Hidden Members

### LINKED ISSUE #27095: DataAnnotationsValidator throws AmbiguousMatchException when there are hidden members
### Describe the bug
We have a Blazor server side application that throws AmbiguousMatchException when editing values.

The DataAnnotationsValidator throws an AmbiguousMatchException when it encounters an object that has hidden members.  This exception is thrown when you edit the input and the validator responds to the NotifyFieldChanged.  The validator uses reflection to determine validation attributes.  When the object being validated includes a hidden member, the GetProperty call without including specific binding flags sees both properties.

### To Reproduce
This is a simple razor page.  The model shown here is the minimum required to reproduce the issue.  In our implementation this same OO pattern is inside a library that we do not have the ability to modify.

The exception will occur if you change the value of the InputNumber and tab away.

```
@using Microsoft.AspNetCore.Components.Forms 
@using Microsoft.AspNetCore.Components.Web 
  
<span>OrderID: @model.OrderID</span> 
  
<EditForm Model="model"> 
    <DataAnnotationsValidator></DataAnnotationsValidator> 
    <InputNumber @bind-Value="@model.OrderID"></InputNumber> 
    <ValidationMessage For="@(() => model.OrderID)"></ValidationMessage> 
    <button type="submit">Submit</button> 
</EditForm> 
  
@code{ 
  
    OrderClass model = new OrderClass() { OrderID = 555 }; 
  
    public class OrderBase 
    { 
        public object OrderID { get; set; } 
    } 
  
    public class OrderClass : OrderBase 
    { 
        [System.ComponentModel.DataAnnotations.Range(1, 100)] 
        public new int OrderID { get; set; } 
    } 
} 
```

### Exceptions (if any)
```
Unhandled exception rendering component: Ambiguous match found.
System.Reflection.AmbiguousMatchException: Ambiguous match found.
at System.RuntimeType.GetPropertyImpl(String name, BindingFlags bindingAttr, Binder binder, Type returnType, Type[] types, ParameterModifier[] modifiers)
at System.Type.GetProperty(String name, BindingFlags bindingAttr)
at System.Type.GetProperty(String name)
at Microsoft.AspNetCore.Components.Forms.EditContextDataAnnotationsExtensions.TryGetValidatableProperty(FieldIdentifier& fieldIdentifier, PropertyInfo& propertyInfo)
at Microsoft.AspNetCore.Components.Forms.EditContextDataAnnotationsExtensions.ValidateField(EditContext editContext, ValidationMessageStore messages, FieldIdentifier& fieldIdentifier)
at Microsoft.AspNetCore.Components.Forms.EditContextDataAnnotationsExtensions.<>c_DisplayClass1_0.<AddDataAnnotationsValidation>b_1(Object sender, FieldChangedEventArgs eventArgs)
at Microsoft.AspNetCore.Components.Forms.EditContext.NotifyFieldChanged(FieldIdentifier& fieldIdentifier)
```

### Further technical details
- ASP.NET Core version
AspNetCore 5.0.0-preview.2.20167.3

- Include the output of `dotnet --info`
```
.NET SDK (reflecting any global.json):
 Version:   5.0.100-rc.2.20479.15
 Commit:    da7dfa8840

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.100-rc.2.20479.15\

Host (useful for support):
  Version: 5.0.0-rc.2.20475.5
  Commit:  c5a3f49c88

.NET SDKs installed:
  5.0.100-rc.2.20479.15 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.21 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.21 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.0-rc.2.20475.17 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.21 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.0-rc.2.20475.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.7 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.0-rc.2.20475.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

- The IDE (VS / VS Code/ VS4Mac) you're running on, and it's version
Microsoft Visual Studio Enteprise 2019 Preview
Version 16.8.0 Preview 4.0 

- Sample Project
[BST_293680.zip](https://github.com/dotnet/aspnetcore/files/5417169/BST_293680.zip)

### PR DESCRIPTION
**Pull request by PreethikaSelvam**
### Bug Description
This PR fixes an issue where the Blazor DataAnnotationsValidator throws an AmbiguousMatchException when validating models that contain hidden properties (i.e., derived class properties declared using new to hide base class members).

Fixes [#27095](https://github.com/dotnet/aspnetcore/issues/27095)

### Problem
When a model includes properties with the same name in both base and derived classes, validation fails with: `System.Reflection.AmbiguousMatchException`. This breaks form validation scenarios in Blazor when inheritance with member hiding is used.

### Root Cause
The issue originates from the use of: `Type.GetProperty(string name)`. This reflection call returns multiple matches when both base and derived types define a property with the same name, resulting in ambiguity and an exception.

### Solution Description
Added handling for AmbiguousMatchException. On exception, retried property resolution using:
`BindingFlags.DeclaredOnly`. This ensures that only properties declared on the current (derived) type are considered, avoiding ambiguity

### Output Screenshots
**Before Fix**

https://github.com/user-attachments/assets/e94c53b3-e06a-4104-9db4-8b5b3e566752

**After Fix**

https://github.com/user-attachments/assets/297a0f9a-9d71-455b-98a5-e9987212f2aa

### Testing
Added 3 new unit test cases covering:
- Validation of models with hidden properties
- Ensuring no exception occurs and correct property resolution
- Manual testing performed to validate real-world scenarios
image

### Impact
- Fixes runtime validation failure for inherited models with hidden members
- Improves robustness of DataAnnotationsValidator
- Ensures consistent behavior across reflection-based validation scenarios

### PR Checklist
- [x] Small and focused — Fixes a specific validation issue
- [x] Tests added — 3 new unit tests included
- [x] Code style matches — Follows existing patterns
- [x] Local build passes
- [x] Local tests pass (14/14)
- [x] No breaking changes
- [x] Issue linked (if applicable) — [#27095](https://github.com/dotnet/aspnetcore/issues/27095)
- [ ] Public API updates — Not applicable
- [x] XML documentation

### CODE REVIEW

File: src/Components/Forms/src/EditContextDataAnnotationsExtensions.cs [hunk 1]
```diff
@@ -269,7 +269,14 @@ private static bool TryGetValidatableProperty(in FieldIdentifier fieldIdentifier
             {
                 // DataAnnotations only validates public properties, so that's all we'll look for
                 // If we can't find it, cache 'null' so we don't have to try again next time
-                propertyInfo = cacheKey.ModelType.GetProperty(cacheKey.FieldName);
+                try
+                {
+                    propertyInfo = cacheKey.ModelType.GetProperty(cacheKey.FieldName, BindingFlags.Public | BindingFlags.Instance);
```
Comment (ilonatommy) [inline src/Components/Forms/src/EditContextDataAnnotationsExtensions.cs]: It does work but managing the correct app flow by exceptions is not the preferred way, if other exist.

We could e.g. use [GetProperties](https://learn.microsoft.com/en-us/dotnet/api/system.type.getproperties?view=net-10.0) (plural) that wouldn't throw, checking ambiguity, and picking the right one but most probably we would need to allocate a list and sort it, which is also not ideal.

Another way that could work is to walk the inheritance tree and choosing the "most-derived" member, e.g. like this:
```
var type = cacheKey.ModelType;
while (type != null && type != typeof(object))
{
    propertyInfo = type.GetProperty(cacheKey.FieldName,
        BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly);
    if (propertyInfo != null) break;
    type = type.BaseType;
}
```
but it adds cost to all, even shallow trees.

So far the best one I found would be keeping `GetProperty` but changing the order: first we check if there's no hidden members, then we take case of inherited ones:
```
propertyInfo = cacheKey.ModelType.GetProperty(
    cacheKey.FieldName,
    BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly);

if (propertyInfo == null)
{
    propertyInfo = cacheKey.ModelType.GetProperty(
        cacheKey.FieldName,
        BindingFlags.Public | BindingFlags.Instance | BindingFlags.FlattenHierarchy);
}
```

Could you, please explore such alternative, cheaper possibilities and choose the best one?

File: src/Components/Forms/test/EditContextDataAnnotationsExtensionsTest.cs [hunk 2]
```diff
@@ -170,6 +170,54 @@ public void CanDetachFromEditContext()
         Assert.Empty(editContext.GetValidationMessages());
     }
 
+    [Fact]
+    public void ValidatesHiddenPropertiesWithoutAmbiguousMatchException()
```
Comment (ilonatommy) [inline src/Components/Forms/test/EditContextDataAnnotationsExtensionsTest.cs:174]: Because we changed the code in `TryGetValidatableProperty` in a way that can alter the behavior not only for hidden properties but for broader cases, let's make sure we have full test coverage for that path. I can see 3 cases that could be added to unit tests:

- Inherited property (no hiding) - verifies that a property defined only on a base class still gets validated when accessed through a derived instance (guards against silent regression where `null` is returned and validation is skipped).
- Multi-level inheritance with hiding - e.g. `C : B : A` where `B` hides `A`'s property with `new` (verifies the correct declaration is resolved at each level).
- Hidden property where only the base has attributes - verifies what happens when the derived property carries no validation attributes but the base one does.

Please, take a moment to think through any other edge cases you can identify in `TryGetValidatableProperty`, then add the missing ones.

File: src/Components/Forms/src/EditContextDataAnnotationsExtensions.cs (hunk 1 above)
Comment (PreethikaSelvam) [reply src/Components/Forms/src/EditContextDataAnnotationsExtensions.cs]: Switched to the two-step DeclaredOnly → FlattenHierarchy lookup in TryGetValidatableProperty.

1. DeclaredOnly resolves the hidden-member case directly, avoiding ambiguity and exception-based flow control.
2. FlattenHierarchy preserves support for inherited properties when no matching property is declared on the runtime type.

This avoids the allocations of a GetProperties()-based approach and the additional reflection overhead of walking the inheritance hierarchy.

File: src/Components/Forms/test/EditContextDataAnnotationsExtensionsTest.cs [hunk 3]
```diff
@@ -170,6 +170,110 @@ public void CanDetachFromEditContext()
...
+        var editContext = new EditContext(model);
+        editContext.EnableDataAnnotationsValidation(_serviceProvider);
+
+        var field = new FieldIdentifier(model, "OrderID");
+        editContext.NotifyFieldChanged(field);
+        Assert.Equal(new[] { "OrderID:range" }, editContext.GetValidationMessages());
+    }
+
+    [Fact]
+    public void ValidatesInheritedPropertyFromBaseClass()
```
Comment (ilonatommy) [inline src/Components/Forms/test/EditContextDataAnnotationsExtensionsTest.cs:222]: The name doesn't match what the test is doing. `DerivedModelWithoutShadowing` hides `BaseName` with `new` keyword and adds `[Range]` to the derived declaration. That's the same hiding scenario already covered by the other tests, not a plain inherited property We should change the test logic: have a derived class with no `BaseName` declaration at all, and validate that the base class's `Required` is still enforced.

File: src/Components/Forms/test/EditContextDataAnnotationsExtensionsTest.cs (hunk 2 above)
Comment (PreethikaSelvam) [reply src/Components/Forms/test/EditContextDataAnnotationsExtensionsTest.cs:174]: Added test coverage for the DeclaredOnly → FlattenHierarchy lookup change, covering hidden properties, multi-level inheritance with property hiding, and cases where validation attributes exist only on the base member. Also added an edge-case test for static properties to help ensure the lookup behavior remains correct and guarded against regressions.

File: src/Components/Forms/test/EditContextDataAnnotationsExtensionsTest.cs (hunk 3 above)
Comment (PreethikaSelvam) [reply src/Components/Forms/test/EditContextDataAnnotationsExtensionsTest.cs:222]: Removed ValidatesInheritedPropertyFromBaseClass, as it was effectively duplicating ValidatesHiddenPropertiesWithoutAmbiguousMatchException.

On the "derived with no BaseName declaration" scenario you described: a test for that will not fail when the source is reverted. By definition, the original `GetProperty(name)` also handles the no-shadowing case correctly (no ambiguity, returns the base property directly), so a test for it would pass on both original and fixed source. The `FlattenHierarchy` fallback path is still exercised indirectly by the existing `IgnoresFieldChangesThatDoNotCorrespondToAValidatableProperty` theory, which uses a flat `TestModel` and goes through the same fallback chain when the field name doesn't match.

Comment (ilonatommy) [reply src/Components/Forms/test/EditContextDataAnnotationsExtensionsTest.cs:222]: > a test for that cannot satisfy the standing rule that every new test must fail when the source is reverted

The point of that test is not to catch a regression in comparison to the original code but to guard against a regression introduced by this very fix.  If a future change to `TryGetValidatableProperty` accidentally breaks plain inherited properties, there would be nothing in the suite to catch it.

The role of tests is not only prove the current fix is justified (fails without the fix but passes with the fix) but also to document the expected behavior. That test should have been in the codebase from the very beginning, it's a coverage gap that we should patch touching this part of the code.

**Review by ilonatommy — APPROVED** [review]

File: src/Components/Forms/test/EditContextDataAnnotationsExtensionsTest.cs (hunk 3 above)
Comment (PreethikaSelvam) [reply src/Components/Forms/test/EditContextDataAnnotationsExtensionsTest.cs:222]: Added ValidatesInheritedPropertyFromBaseClass as a regression guard for the DeclaredOnly → FlattenHierarchy lookup path. It uses a derived type that does not redeclare BaseName and verifies that validation is still applied from the base property's [Required] attribute.

**Comment by ilonatommy** [comment]
The CI is always running on PR branch merged to main. In case there would be unresolved conflicts, there would be an information about it on the PR. Merging with main when there are no conflicts and when you're not aware of any recent changes that could influence your feature prolongs the time-to-merge of the PR.

**Review by ilonatommy — APPROVED** [review]

File: src/Components/Forms/src/EditContextDataAnnotationsExtensions.cs [hunk 4]
```diff
@@ -367,7 +367,16 @@ private static bool TryGetValidatableProperty(in FieldIdentifier fieldIdentifier
...
-                propertyInfo = cacheKey.ModelType.GetProperty(cacheKey.FieldName);
+                propertyInfo = cacheKey.ModelType.GetProperty(
+                    cacheKey.FieldName,
+                    BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly);
+
+                if (propertyInfo is null)
+                {
+                    propertyInfo = cacheKey.ModelType.GetProperty(
+                        cacheKey.FieldName,
+                        BindingFlags.Public | BindingFlags.Instance | BindingFlags.FlattenHierarchy);
```
Comment (Youssef1313) [inline src/Components/Forms/src/EditContextDataAnnotationsExtensions.cs:378]: FlattenHierarchy is only relevant for statics IIRC.

And this code can still throw if the previous call returned null and we get into here with some shadowing member in a base type.
